### PR TITLE
Add Cow protocol price improvement data

### PR DESCRIPTION
**Pull request by josojo**
Replaces: https://github.com/duneanalytics/spellbook/pull/2478. Most of the work was done by @gentrexha !

**For Dune Engine V2**
The results generated by this query are very similar to the ones of the old v1 version:
Histogram slippage
v1: https://dune.com/queries/1039434/1792834
v2: https://dune.com/queries/1901957
Sum daily slippage:
v1: https://dune.com/queries/1035637/1786462
v2: https://dune.com/queries/1902617
I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory) See here: https://dune.com/queries/1902930?d=1
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [x] `coin_id` represents the ID of the coin on coinpaprika.com
* [x] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

File: models/cow_protocol/ethereum/cow_protocol_ethereum_price_improvement.sql [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{{ config(alias='price_improvement',
+        post_hook='{{ expose_spells_hide_trino(\'["ethereum"]\',
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_price_improvement.sql:2]: Do you explicitly want to make this unavailable in duneSQL?

File: models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml [hunk 2]
```diff
@@ -273,4 +273,47 @@ models:
...
+      - *block_time
+      - &order_type
+        name: order_type
+      - *limit_buy_amount
+      - &buy_quote
+        name: buy_quote
+      - *atoms_bought
+      - *units_bought
+      - &buy_decimals
+        name: buy_decimals
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml:297]: I really don't think decimals are necessary to keep around here.

File: models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml [hunk 3]
```diff
@@ -273,4 +273,47 @@ models:
...
+      is called price improvement and is measured as percentage or in absolte dollars.
+    columns:
+      - *block_time
+      - &order_type
+        name: order_type
+      - *limit_buy_amount
+      - &buy_quote
+        name: buy_quote
+      - *atoms_bought
+      - *units_bought
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml:295]: These are already part of the trades table. Why keep them here?

File: models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml [hunk 4]
```diff
@@ -273,4 +273,47 @@ models:
...
+      blockchain: ethereum
+      project: cow_protocol
+      contributors: bh2smith, gentrexha, josojo
+    config:
+      tags: [ 'ethereum','cow_protocol','price improvement' ]
+    description: >
+      The query contains several measurments between the quoted price on cowswap and 
+      the executed price. The price difference between the quote and the actual price 
+      is called price improvement and is measured as percentage or in absolte dollars.
+    columns:
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml:287]: Missing block number! Please include.

**Review by bh2smith — APPROVED** [review]
Looks good but missing block number which should be included. 

I think there are a bunch of unnecessary columns in this table that can easily be acquired with a join on trades.  Specifically I would suggest removing prices, trade value, token decimals amounts and atoms bought/sold. Thus table should contain information on top of what is already in the trades table.

**Review by bh2smith — CHANGES_REQUESTED** [review]
Add block number

**Comment by bh2smith** [comment]
How is this coming along @josojo?

File: models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml [hunk 5]
```diff
@@ -273,4 +273,31 @@ models:
...
+    meta:
+      blockchain: ethereum
+      project: cow_protocol
+      contributors: bh2smith, gentrexha, josojo
+    config:
+      tags: [ 'ethereum','cow_protocol','price improvement' ]
+    description: >
+      The query contains several measurments between the quoted price on cowswap and 
+      the executed price. The price difference between the quote and the actual price 
+      is called price improvement and is measured as percentage or in absolte dollars.
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_schema.yml:286]:
```suggestion
      is called price improvement and is measured as percentage or in absolute dollars.
```

File: models/cow_protocol/ethereum/cow_protocol_ethereum_price_improvement.sql [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            ELSE (((limit_sell_amount / sell_tolerance) - (atoms_sold)) / (limit_sell_amount / sell_tolerance))
+    END) AS surplus_percentage,
+    CASE
+        WHEN order_type = 'SELL'
+        THEN (atoms_bought - buy_quote) / POWER(10, buy_decimals) * (usd_value / units_bought)
+        ELSE ((limit_sell_amount / sell_tolerance) - (atoms_sold)) * (usd_value / (limit_sell_amount / sell_tolerance))
+      END AS surplus_usd
+    from raw_data
+)
+select * from results
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_price_improvement.sql:56]:
```suggestion
select * from results

```

**Review by bh2smith — APPROVED** [review]
Looks pretty good to me! I would say its ready for merge!

**Review by soispoke — APPROVED** [review]
lgtm!

**Pull request merged**